cfn/tiler_api_stack.py

```diff
@@ -11,242 +14,379 @@
+        'InstanceSecurityGroup',
+        GroupDescription='Enable SSH access via port 22',
+        SecurityGroupIngress=[
+            ec2.SecurityGroupRule(
+                IpProtocol='tcp',
+                FromPort='22',
+                ToPort='22',
+                CidrIp=ALLOW_ALL_CIDR),
+            ec2.SecurityGroupRule(
+                IpProtocol='tcp',
+                FromPort='80',
+                ToPort='80',
+                CidrIp=ALLOW_ALL_CIDR)],
+        VpcId=Ref(VPC),
+    ))
 
 #
-# Auto Scaling Group Resources
+# Instance
 #
-tiler_launch_config = t.add_resource(asg.LaunchConfiguration(
-    'lcTiler',
-    ImageId=Ref(tiler_ami_param),
-    # IamInstanceProfile=Ref(tiler_instance_profile_param),
-    InstanceType=Ref(tiler_instance_type_param),
-    KeyName=Ref(keyname_param),
-    SecurityGroups=[Ref(tiler_security_group)],
-    UserData=Base64(read_file('cloud-config/oam-server-api.yml'))
-))
 
-tiler_auto_scaling_group = t.add_resource(asg.AutoScalingGroup(
-    'asgTiler',
-    AvailabilityZones=Ref(availability_zones_param),
-    Cooldown=300,
-    DesiredCapacity=1,
-    HealthCheckGracePeriod=600,
-    HealthCheckType='ELB',
-    LaunchConfigurationName=Ref(tiler_launch_config),
-    LoadBalancerNames=[Ref(tiler_load_balancer)],
-    MaxSize=10,
-    MinSize=1,
-    NotificationConfigurations=[asg.NotificationConfigurations(
-        TopicARN=Ref(notification_arn_param),
-        NotificationTypes=[
-            asg.EC2_INSTANCE_LAUNCH,
-            asg.EC2_INSTANCE_LAUNCH_ERROR,
-            asg.EC2_INSTANCE_TERMINATE,
-            asg.EC2_INSTANCE_TERMINATE_ERROR
-        ]
-    )],
-    VPCZoneIdentifier=Ref(private_subnets_param),
-    Tags=[asg.Tag('Name', 'Tiler', True)]
+refs = dict(
+    aws_stack_name=Ref('AWS::StackName'),
+    aws_region=Ref('AWS::Region'),
+    keyname=Ref(keyname_param),
+    server_api_version=Ref(server_api_version_param),
+    publisher_version=Ref(publisher_version_param),
+    small_cluster_size=Ref(small_cluster_size_param),
+    medium_cluster_size=Ref(medium_cluster_size_param),
+    large_cluster_size=Ref(large_cluster_size_param),
+    small_image_count=Ref(small_threshold_param),
+    medium_image_count=Ref(medium_threshold_param),
+    emr_master_instance_type=Ref(emr_master_instance_type_param),
+    emr_core_cluster_size=Ref(emr_core_cluster_size_param),
+    emr_core_instance_type=Ref(emr_core_instance_type_param),
+    emr_task_instance_type=Ref(emr_task_instance_type_param),
+    emr_task_bid_price=Ref(emr_task_bid_price_param),
+    auth_token_bucket=Ref(auth_token_bucket_param),
+    auth_token_key=Ref(auth_token_key_param),
+    status_bucket=Ref(status_bucket_param),
+    status_prefix=Ref(status_prefix_param),
+    status_sqs_queue_url=Ref(status_sqs_queue_url_param),
+    oam_api_token=Ref(oam_api_token_param),
+)
+
+# convert to JSON representations
+refs = dict([(k, '|||' + json.dumps(v.JSONrepr()) + '|||') for (k, v) in refs.items()])
```

Comment:
I think that I would have ended up doing something similar given the number of variables being passed in. Only thing additional I would have offered would have been to use something like Jinja2 to do the templating, but this seems to be working fine.